File: 39-Combination Sum-组合总和/39-Combination Sum-组合总和.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
 let res = []
 let path = []
 let sum = 0
 var combinationSum = function(candidates, target) {
     res = []
     backtrace(candidates, target, 0)
     return res
 };
 
 var backtrace = function(candidates, target, index){
     // 关键，终止条件
     if(sum > target){
         return
     }
     if(sum === target){
         res.push(Array.from(path))
         return
     }
 
     for(let i = index; i <  candidates.length; i++){
         path.push(candidates[i])
         sum += candidates[i]
         backtrace(candidates, target, i)
         path.pop()
         sum -= candidates[i]
     }
 }

File: 39-Combination Sum-组合总和/39-Combination Sum-组合总和.java
class Solution {
    public List<List<Integer>> res;
    public int sum;
    public LinkedList<Integer> path;

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        res = new ArrayList<>();
        path = new LinkedList<Integer>();
        backtrace(candidates, target, 0);
        return res;
    }

    public void backtrace(int[] candidates, int target, int index){
        if(sum > target){
            return;
        }
        if(sum == target){
            res.add(new ArrayList<Integer>(){{addAll(path);}});
            return;
        }

        for(int i = index; i < candidates.length; i++){
            path.add(candidates[i]);
            sum += candidates[i];
            backtrace(candidates, target, i);
            path.removeLast();
            sum -= candidates[i];
        }
    }
}